### PR TITLE
source_code_interpretation/plots_py.md

### PR DESCRIPTION
**Pull request by ccssu**

### CODE REVIEW

File: docs/source_code_interpretation/utils/plots_py.md [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+    xywh2xyxy,
+    xyxy2xywh,
+)
+from utils.metrics import fitness
+
+# Settings
+# 设置一些基本的配置  Settings
+RANK = -1  # int(os.getenv('RANK', -1))
+matplotlib.rc("font", **{"size": 11})
+# 如果这句话放在import matplotlib.pyplot as plt之前就算加上plt.show()也不会再屏幕上绘图 放在之后其实没什么用
```
Comment (BBuf) [inline docs/source_code_interpretation/utils/plots_py.md:58]: 这句话有点奇怪

File: docs/source_code_interpretation/utils/plots_py.md [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+        return (c[2], c[1], c[0]) if bgr else c
+
+    @staticmethod
+    def hex2rgb(h):  # rgb order (PIL)
+        return tuple(int(h[1 + i : 1 + i + 2], 16) for i in (0, 2, 4))
+
+
+# 初始化Colors对象 下面调用colors的时候会调用 __call__函数 (https://www.jianshu.com/p/33d6688246a6)
+# __call__: https://docs.python.org/3/reference/datamodel.html?highlight=__call__#emulating-callable-objects
+colors = Colors()  # create instance for 'from utils.plots import colors'
```
Comment (BBuf) [inline docs/source_code_interpretation/utils/plots_py.md:104]:
```suggestion
colors = Colors()  # 为 'from utils.plots import colors' 创建实例
```

File: docs/source_code_interpretation/utils/plots_py.md [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+
+## 2.save_one_box
+> 将预测到的目标从原图中扣出来
+
+
+```python
+def save_one_box(
+    xyxy, im, file=Path("im.jpg"), gain=1.02, pad=10, square=False, BGR=False, save=True
+):
+    """用在detect.py文件中  由opt的save-crop参数控制 执不执行
```
Comment (BBuf) [inline docs/source_code_interpretation/utils/plots_py.md:115]:
```suggestion
    """用在detect.py文件中  由opt的save-crop参数控制是否执行
```

File: docs/source_code_interpretation/utils/plots_py.md [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+    将预测到的目标从原图中扣出来 剪切好 并保存 会在 runs/detect/expn 下生成crops文件, 将剪切的图片保存在里面
+    Save image crop as {file} with crop size multiple {gain} and {pad} pixels. Save and/or return crop
+    :params xyxy: 预测到的目标框信息 list 4个tensor x1 y1 x2 y2 左上角 + 右下角
+    :params im: 原图片 需要裁剪的框从这个原图上裁剪  nparray  (1080, 810, 3)
+    :params file: runs\detect\exp\crops\dog\bus.jpg
+    :params gain: 1.02 xyxy缩放因子
+    :params pad: xyxy pad一点点边界框 裁剪出来会更好看
+    :params square: 是否需要将xyxy放缩成正方形
```
Comment (BBuf) [inline docs/source_code_interpretation/utils/plots_py.md:123]: 什么叫把xyxy放缩成正方形，可以解释一下

File: docs/source_code_interpretation/utils/plots_py.md [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+    :params BGR: 保存的图片是BGR还是RGB 。(什么是RGB模式与BGR模式请参阅读 https://blog.csdn.net/SGchi/article/details/104474976)
+    :params save: 是否要保存剪切的目标框
+    """
+    # Save image crop as {file} with crop size multiple {gain} and {pad} pixels. Save and/or return crop
+    # view函数: Returns a new tensor with the same data as the self tensor but of a different shape.
+    # https://pytorch.org/docs/stable/generated/torch.Tensor.view.html?highlight=torch+view#torch.Tensor.view
+    # 其实按照我们通俗的理解就是reshape，
+    # 只不过这里是reshape的是张量，也就是将张量重新调整为自己想要的维度（形状大小）
+    xyxy = flow.tensor(xyxy).view(-1, 4)  # list -> Tensor [1, 4] = [x1 y1 x2 y2]
```
Comment (BBuf) [inline docs/source_code_interpretation/utils/plots_py.md:132]: 这个tensor的维度是[1, 4]还是[bs, 4]，意思是这个函数的输入只有一个bbox还是可以有很多个bbox呢？

File: docs/source_code_interpretation/utils/plots_py.md [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+    :params save_dir: runs\train\exp21
+    :params loggers: 日志对象
+    """
+    LOGGER.info(f"Plotting labels to {save_dir / 'labels.jpg'}... ")
+    c, b = labels[:, 0], labels[:, 1:].transpose()  # classes, boxes
+    nc = int(c.max() + 1)  # number of classes
+    # pd.DataFrame: 创建DataFrame, 类似于一种excel, 表头是['x', 'y', 'width', 'height']  表格数据: b中数据按行依次存储
+    x = pd.DataFrame(b.transpose(), columns=["x", "y", "width", "height"])
+
+    # seaborn correlogram
```
Comment (BBuf) [inline docs/source_code_interpretation/utils/plots_py.md:236]: 这一节我感觉可以单独做一节教程，就是我之前说过的那个：

2是解析一下one-yolov5训练的时候，train日志里面每张图代表什么以及细节（比如横纵坐标以及计算方法）

Comment (BBuf) [reply docs/source_code_interpretation/utils/plots_py.md:236]: 因为这里的代码有点散，很难看得出来下去，如果结合每张图是用哪行画出来的方式就很容易接受了
